### PR TITLE
Nelz/cobra

### PR DESCRIPTION
**Pull request by nelz9999**
This is a fairly invasive structural change. Uses cobra to build the commands, and allows for inputs either via envvars or via command-line flags.

Also, most of the sub-funcs should be a tiny bit more unit-test-able, even tho I haven't done that here yet.

This was most of the "heavy-lifting", so about 80% of the work. The remaining bits of bug hunting and unit testing should only be about another 80%. :trollface: 

This is currently in PR mostly for gathering feedback.

### CODE REVIEW

**Comment by nelz9999** [comment]
```$ ./buildevents --help

The buildevents executable creates Honeycomb events and tracing information
about your Continuous Integration builds.

Usage:
  buildevents [command]

Available Commands:
  build       Sends the root span for the entire build
  cmd         Invoke an individual command that is part of the build.
  help        Help about any command
  step        Joins a collection of individual commands
  watch       Polls the CircleCI API and waits until all jobs have finished.

Flags:
  -a, --apihost string    [env.BUILDEVENT_APIHOST] the hostname for the Honeycomb API server to which to send this event (default "https://api.honeycomb.io")
  -k, --apikey string     [env.BUILDEVENT_APIKEY] the Honeycomb authentication token
  -d, --dataset string    [env.BUILDEVENT_DATASET] the name of the Honeycomb dataset to which to send these events (default "buildevents")
  -f, --filename string   [env.BUILDEVENT_FILE] the path of a text file holding arbitrary key=val pairs (multi-line-capable, logfmt style) to be added to the Honeycomb event
  -h, --help              help for buildevents
  -p, --provider string   [env.BUILDEVENT_CIPROVIDER] if unset, will inspect the environment to try and detect Travis-CI, CircleCI, or GitLab-CI.
      --version           version for buildevents

Use "buildevents [command] --help" for more information about a command.```

**Comment by nelz9999** [comment]
```$ ./buildevents watch --help

Polls the CircleCI API and waits until all jobs have finished (either
succeeded, failed, or are blocked). It then reports the final status of the
build with the appropriate timers.

Usage:
  buildevents watch BUILD_ID [flags]

Flags:
  -c, --circlekey string    [env.BUILDEVENT_CIRCLE_API_TOKEN] CircleCI API token used for watching builds for private repositories
  -h, --help                help for watch
  -j, --jobname string      [env.CIRCLE_JOB] CircleCI identifier for the current job
  -t, --timeout int         [env.BUILDEVENT_TIMEOUT] maximum time (in minutes) that watch should wait before timing out (default 10)
  -w, --workflowid string   [env.CIRCLE_WORKFLOW_ID] CircleCI identifier for the current workflow

Global Flags:
  -a, --apihost string    [env.BUILDEVENT_APIHOST] the hostname for the Honeycomb API server to which to send this event (default "https://api.honeycomb.io")
  -k, --apikey string     [env.BUILDEVENT_APIKEY] the Honeycomb authentication token
  -d, --dataset string    [env.BUILDEVENT_DATASET] the name of the Honeycomb dataset to which to send these events (default "buildevents")
  -f, --filename string   [env.BUILDEVENT_FILE] the path of a text file holding arbitrary key=val pairs (multi-line-capable, logfmt style) to be added to the Honeycomb event
  -p, --provider string   [env.BUILDEVENT_CIPROVIDER] if unset, will inspect the environment to try and detect Travis-CI, CircleCI, or GitLab-CI.```

**Comment by nelz9999** [comment]
Currently, if there's no APIKey given, the executable will output what _would_ get sent...

```
$ ./buildevents step bubba gump 1561844684 fiction
{"data":{"duration_ms":84464113,"meta.version":"dev","name":"fiction","service_name":"step","trace.parent_id":"bubba","trace.span_id":"gump","trace.trace_id":"bubba"},"time":"2019-06-29T14:44:44-07:00","dataset":"buildevents"}```

**Comment by nelz9999** [comment]
FYI, because the changes were so structural, i don't feel like the GH side-by-side view adds anything... I'd suggest just looking at the raw files themselves to get a better understanding.

**Comment by maplebed** [comment]
structural - it'd be a huge help to pull each command out of an inline function and into its own function, perhaps its own file. Or maybe all in one (separate) file - anything to get the logic of each command separated from a gargantuan `main()`

File: main.go [hunk 1]
```diff
@@ -224,161 +119,355 @@ func addEnvVars(ciProvider string) {
...
+			ev := createEvent(config, ciProvider, args[0])
+			defer ev.Send()
+
+			providerInfo(ciProvider, ev)
+			arbitraryFields(filename, ev)
+
+			ts, dur := parseSince(args[1])
+			ev.Add(map[string]interface{}{
+				"service_name":  "build",
+				"trace.span_id": args[0],
```
Comment (maplebed) [inline main.go]: I would really like to get rid of `arg[#]` parsing buried in these spots. I can understand not wanting to make the change away from positional arguments right now. Could you move these to the head of the function at least, so it's clear what is coming in? I also see something in the Cobra docs about positional arguments. Perhaps we can leverage that to make this a little cleaner.

**Comment by nelz9999** [comment]
Cool cool. Like I said, I mostly got this to "solicit feedback" level. Now that I've gotten a positive response, I can do more fine-tuning of the code. 👍

File: main.go [hunk 2]
```diff
@@ -224,161 +119,355 @@ func addEnvVars(ciProvider string) {
...
-		if _, present := os.LookupEnv("TRAVIS"); present {
-			ciProvider = "Travis-CI"
-		} else if _, present := os.LookupEnv("CIRCLECI"); present {
-			ciProvider = "CircleCI"
-		} else if _, present := os.LookupEnv("GITLAB_CI"); present {
-			ciProvider = "GitLab-CI"
-		}
+	defer libhoney.Close()
+	config := libhoney.Config{}
+	filename := ""
```
[the remark this answers is not in the document]
Comment (nelz9999) [reply main.go]: Yupperz

File: main.go [hunk 3]
```diff
@@ -224,161 +119,355 @@ func addEnvVars(ciProvider string) {
...
+		Short:   "buildevents creates events for your CI builds",
+		Long: `
+The buildevents executable creates Honeycomb events and tracing information
+about your Continuous Integration builds.`,
 	}
 
-	// use defaults for dataset and apihost if they're unset.
-	if dataset == "" {
-		dataset = "buildevents"
+	root.PersistentFlags().StringVarP(&config.APIKey, "apikey", "k", "", "[env.BUILDEVENT_APIKEY] the Honeycomb authentication token")
```
[the remark this answers is not in the document]
Comment (nelz9999) [reply main.go]: Will do!

File: main.go [hunk 4]
```diff
@@ -224,161 +119,355 @@ func addEnvVars(ciProvider string) {
...
-		fmt.Fprintf(os.Stderr, "problems loading from %q: %v\n", locn, err)
+		fmt.Fprintf(os.Stderr, "problems loading from %q: %v\n", loc, err)
 	}
 }
 
-func usage() {
-	fmt.Printf(`Usage: buildevents [build,step,cmd] ... args
-
-	For documentation, see https://github.com/honeycombio/buildevents
+func parseSince(start string) (time.Time, time.Duration) {
```
[the remark this answers is not in the document]
Comment (nelz9999) [reply main.go]: Gonna clean this up a bit.

File: main.go (hunk 1 above)
Comment (nelz9999) [reply main.go]: Cleaning this up!

**Comment by nelz9999** [comment]
@maplebed I've moved stuff around a bit, and did some little tidy and habitability improvements. 

My testing was fairly lightweight... I'm wondering if you can try the executable in a more real-world way.

This is the kind of stuff I was doing to test:
```
$ echo "gooooob=morning" >> example.txt
$ ./buildevents -d yahoolie -f example.txt cmd BuildMe StepMe ecccccccco -- echo fungusamongus
running /bin/bash -c "echo" "fungusamongus"
fungusamongus
{"data":{"cmd":"\"echo\" \"fungusamongus\"","duration_ms":7,"gooooob":"morning","meta.version":"dev","name":"ecccccccco","service_name":"cmd","status":"success","trace.parent_id":"StepMe","trace.span_id":"8ce1344ee38b58f05f62cfcdde735781","trace.trace_id":"BuildMe"},"time":"2019-07-03T12:22:34.988203-07:00","dataset":"yahoolie"}
```

**Comment by maplebed** [comment]
I ran it through its paces and things look good. I caught one regression (lost the `strings.TrimSpace()` on all arguments with the refactor) that I'll fix in a followup PR.

Thanks for this PR! It'll make future buildevents work much more fun.

**Pull request merged**